File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        for(int i=0;i<nums.size();i++){
            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }

            int target = -nums[i];
            int left = i+1;
            int right = nums.size()-1;

            while(left < right){
                int sum = nums[left]+ nums[right];
                if( sum == target){
                    result.push_back({nums[left], nums[right], nums[i]});
                
                while(left< right && nums[left]==nums[left+1]){
                    ++left;
                }
                while(left< right && nums[right]==nums[right-1]){
                    --right;
                }
                ++left;
                --right;
                }
                else if(sum< target){
                    left++;
                }
                else{
                    --right;
                }
            }
        }
        return result;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        result = []
        k=0
        nums.sort()
        for k in range(len(nums)-2):
            if k > 0 and nums[k] == nums[k - 1]:
                continue
            target = nums[k]
            left = k+1         
            right = len(nums)- 1
            while(left < right):
                if(nums[left] + nums[right] > -target):
                    right-=1
                elif(nums[left] + nums[right] < -target):
                    left+=1
                else:
                    result.append([target, nums[left], nums[right]])
                    left+=1
                    right-=1
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1

                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1
        return result

                
        